### PR TITLE
Coverage providers should add coverage

### PR DESCRIPTION
**Pull request by leonardr**
This branch fixes two problems with coverage records not being added:

* When ensure_coverage() checked for an existing coverage record, it didn't take the operation into account. So if your coverage provider ran operation 'foo' and there was an existing record for operation 'bar', it would seem as though the 'foo' operation had completed. No work would be done and no new coverage record would be created.

* The run_on_identifiers() method was doing the work, but not creating coverage records reflecting the work done. This is because it called process_batch(), which does the work, but not running the coverage record code, which was duplicated across run_once() and ensure_coverage(). I refactored this code into a new method, process_batch_and_handle_results(). I also added tests for the code in process_batch_and_handle_results, which was previously untested.

I normalized the return value of all of these methods so they return a 2-tuple (counts, records), where `counts` is a 3-tuple (# of successes, # of transient failures, # of persistent failures) and `records` is a mixed list of `CoverageRecord` and `CoverageFailure` objects.

I tested a previously untested code path using a CoverageProvider that simply ignores all the work given to it. (This is treated as a transient failure.)

### CODE REVIEW

File: coverage.py [hunk 1]
```diff
@@ -122,49 +135,77 @@ def run_once(self, offset):
...
+        )
+
+        # Ignore transient failures.
+        return offset + transient_failures
 
+    def process_batch_and_handle_results(self, batch):
+        """:return: A 2-tuple (counts, records). 
+
+        `counts` is a 3-tuple (successes, transient failures,
+        persistent_failures).
```
Comment (aslagle) [inline coverage.py]: be consistent with calling it persistent vs permanent throughout this code

File: monitor.py [hunk 2]
```diff
@@ -155,17 +155,17 @@ def resolve(self, unresolved_identifier):
         for provider in self.required_coverage_providers:
             if not identifier.type in provider.input_identifier_types:
                 continue
-            record = provider.ensure_coverage(identifier, force=True)
-            if record.exception:
-                raise ResolutionFailed(500, record.exception)                
+            counts, records = provider.ensure_coverage(identifier, force=True)
+            if records and records[0].exception:
+                raise ResolutionFailed(500, records[0].exception)
```
Comment (aslagle) [inline monitor.py]: this needs to be changed back

File: monitor.py [hunk 3]
```diff
@@ -155,17 +155,17 @@ def resolve(self, unresolved_identifier):
...
 
         # Now go through the optional providers. It's the same deal,
         # but a CoverageFailure doesn't cause the entire identifier
         # resolution process to fail.
         for provider in self.optional_coverage_providers:
             if not identifier.type in provider.input_identifier_types:
                 continue
-            record = provider.ensure_coverage(identifier, force=True)
+            counts, records = provider.ensure_coverage(identifier, force=True)
```
Comment (aslagle) [inline monitor.py]: this too

File: monitor.py (hunk 2 above)
Comment (leonardr) [inline monitor.py]: Sorry that this is a moving target. :sweat_smile:

**Comment by leonardr** [comment]
OK, I've changed ensure_coverage() so that it always returns either a CoverageRecord or a CoverageFailure, even if it ends up not doing any work. This should remove the need to make changes in other branches, and I think it also removes some bugs where some code was really expecting to get something back from ensure_coverage(), and receiving None would have crashed it.

**Comment by aslagle** [comment]
:+1:

**Pull request merged**